crates/bevy_ecs/src/world/interior_mutable_world.rs

```diff
@@ -0,0 +1,473 @@
+///
+/// This alone is not enough to implement bevy systems where multiple systems can access *disjoint* parts of the world concurrently. For this, bevy stores all values of
+/// resources and components (and [`ComponentTicks`](crate::component::ComponentTicks)) in [`UnsafeCell`](std::cell::UnsafeCell)s, and carefully validates disjoint access patterns using
+/// APIs like [`System::component_access`](crate::system::System::component_access).
+///
+/// A system then can be executed using [`System::run_unsafe`](crate::system::System::run_unsafe) with a `&World` and use methods with interior mutability to access resource values.
+/// access resource values.
+///
+/// ### Example Usage
+///
+/// [`InteriorMutableWorld`] can be used as a building block for writing APIs that safely allow disjoint access into the world.
+/// In the following example, the world is split into a resource access half and a component access half, where each one can
+/// safely hand out mutable references.
+///
+/// ```
+/// use bevy_ecs::world::World;
+/// use bevy_ecs::change_detection::Mut;
+/// use bevy_ecs::system::Resource;
+/// use bevy_ecs::world::interior_mutable_world::InteriorMutableWorld;
+///
+/// // INVARIANT: existance of this struct means that users of it are the only ones being able to access resources in the world
+/// struct OnlyResourceAccessWorld<'w>(InteriorMutableWorld<'w>);
+/// // INVARIANT: existance of this struct means that users of it are the only ones being able to access components in the world
+/// struct OnlyComponentAccessWorld<'w>(InteriorMutableWorld<'w>);
+///
+/// impl<'w> OnlyResourceAccessWorld<'w> {
+///     fn get_resource_mut<T: Resource>(&mut self) -> Option<Mut<'w, T>> {
+///         // SAFETY: resource access is allowed through this InteriorMutableWorld
+///         unsafe { self.0.get_resource_mut::<T>() }
+///     }
+/// }
+/// // impl<'w> OnlyComponentAccessWorld<'w> {
+/// //     ...
+/// // }
+///
+/// // the two interior mutable worlds borrow from the `&mut World`, so it cannot be accessed while they are live
+/// fn split_world_access(world: &mut World) -> (OnlyResourceAccessWorld<'_>, OnlyComponentAccessWorld<'_>) {
+///     let interior_mutable_world = world.as_interior_mutable();
+///     let resource_access = OnlyResourceAccessWorld(interior_mutable_world);
+///     let component_access = OnlyComponentAccessWorld(interior_mutable_world);
+///     (resource_access, component_access)
+/// }
+/// ```
+#[derive(Copy, Clone)]
+pub struct InteriorMutableWorld<'w>(&'w World);
+
+impl<'w> InteriorMutableWorld<'w> {
+    pub(crate) fn new(world: &'w World) -> Self {
+        InteriorMutableWorld(world)
+    }
+
+    /// Gets a reference to the [`&World`](crate::world::World) this [`InteriorMutableWorld`] belongs to.
+    /// This can be used to call methods like [`World::read_change_tick`] which aren't exposed here but don't perform any accesses.
```

Comment:
Can you add a `InteriorMutableWorld::change_tick` method in addition to this one?